src/Lykke.AlgoStore.Services/Validation/CSharpCodeBuildSession.cs

```diff
@@ -0,0 +1,309 @@
+
+            _compilation = _compilation.WithOptions(new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
+
+            _semanticModel = _compilation.GetSemanticModel(_syntaxTree, false);
+            var semanticDiagnostics = _semanticModel.GetDiagnostics();
+
+            validationMessages.AddRange(semanticDiagnostics.Select(DiagnosticToValidationMessage));
+
+            return CreateAndSetValidationResult(out _syntaxValidationResult, 
+                                                !ErrorExists(validationMessages), 
+                                                validationMessages);
+        }
+
+        public async Task<AlgoMetaDataInformation> ExtractMetadata()
+        {
+            var root = (CompilationUnitSyntax)await _syntaxTree.GetRootAsync();
+            var namespaceDeclaration = (NamespaceDeclarationSyntax)root.Members[0];
+            var classDeclaration = namespaceDeclaration.Members
+                .OfType<ClassDeclarationSyntax>()
+                .First();
+
+            var newAssembly = GenerateAssembly();
+            var newType = newAssembly.GetType($"{namespaceDeclaration.Name}.{classDeclaration.Identifier.Text}");
+            var metadata = ExtractMetadata(newType);
+
+            return metadata;
+        }
+
+        private AlgoMetaDataInformation ExtractMetadata(Type algoType)
+        {
+            var metadata = new AlgoMetaDataInformation
+            {
+                Functions = new List<AlgoMetaDataFunction>(),
+                Parameters = new List<AlgoMetaDataParameter>()
+            };
+
+            var algoProperties = algoType
+                .GetProperties(BindingFlags.Instance | BindingFlags.Public);
+
+            foreach (var algoProperty in algoProperties)
+            {
+                var parameter = ToAlgoMetaDataParameter(algoProperty);
+
+                if (algoProperty.PropertyType.IsEnum)
+                    parameter.PredefinedValues = ToEnumValues(algoProperty);
+
+                metadata.Parameters.Add(parameter);
+            }
+
+            var functionProperties = algoType
+                .GetRuntimeFields()
+                .Where(x => x.FieldType.BaseType == typeof(AbstractFunction));
+
+            foreach (var functionProperty in functionProperties)
+            {
+                var function = ToAlgoMetadataFunction(functionProperty);
+
+                var functionParamBasePropertyField = functionProperty
+                    .FieldType
+                    .GetFields()
+                    .FirstOrDefault(x => x.FieldType.BaseType == typeof(FunctionParamsBase));
+
+                //There is a public property which base type is of FunctionParamBase type
+                if (functionParamBasePropertyField != null)
+                {
+                    function.FunctionParameterType = functionParamBasePropertyField.FieldType.FullName;
+                    var functionParameters = functionParamBasePropertyField.FieldType.GetProperties();
```

Comment:
`GetProperties()` includes static properties in the search. Also, must check if property has a setter